rct229/rulesets/ashrae9012019/section4/section4rule2.py

```diff
@@ -0,0 +1,269 @@
+                    "maximum_humidity_stpt_sch_id_b": maximum_humidity_stpt_sch_id_b,
+                    "maximum_humidity_stpt_sch_id_p": maximum_humidity_stpt_sch_id_p,
+                    "minimum_humidity_schedule_not_matched": minimum_humidity_schedule_not_matched,
+                    "maximum_humidity_schedule_not_matched": maximum_humidity_schedule_not_matched,
+                    "minimum_humidity_schedule_matched": minimum_humidity_schedule_matched,
+                    "maximum_humidity_schedule_matched": maximum_humidity_schedule_matched,
+                }
+
+            def manual_check_required(self, context, calc_vals=None, data=None):
+                minimum_humidity_schedule_not_matched = calc_vals[
+                    "minimum_humidity_schedule_not_matched"
+                ]
+                maximum_humidity_schedule_not_matched = calc_vals[
+                    "maximum_humidity_schedule_not_matched"
+                ]
+                return (
+                    minimum_humidity_schedule_not_matched
+                    or maximum_humidity_schedule_not_matched
+                )
+
+            def rule_check(self, context, calc_vals=None, data=None):
+                minimum_humidity_schedule_matched = calc_vals[
+                    "minimum_humidity_schedule_matched"
+                ]
+                maximum_humidity_schedule_matched = calc_vals[
+                    "maximum_humidity_schedule_matched"
+                ]
+                return (
+                    minimum_humidity_schedule_matched
+                    and maximum_humidity_schedule_matched
+                )
+
+            def get_fail_msg(self, context, calc_vals=None, data=None):
+                minimum_humidity_stpt_sch_id_b = calc_vals[
+                    "minimum_humidity_stpt_sch_id_b"
+                ]
+                minimum_humidity_stpt_sch_id_p = calc_vals[
+                    "minimum_humidity_stpt_sch_id_p"
+                ]
+                maximum_humidity_stpt_sch_id_b = calc_vals[
+                    "maximum_humidity_stpt_sch_id_b"
+                ]
+                maximum_humidity_stpt_sch_id_p = calc_vals[
+                    "maximum_humidity_stpt_sch_id_p"
+                ]
+                if (
+                    minimum_humidity_stpt_sch_id_b
+                    and not minimum_humidity_stpt_sch_id_p
+                ) or (
+                    maximum_humidity_stpt_sch_id_b
+                    and not maximum_humidity_stpt_sch_id_p
+                ):
+                    return FAIL_MSG_B
+                elif (
+                    not minimum_humidity_stpt_sch_id_b
+                    and minimum_humidity_stpt_sch_id_p
+                ) or (
+                    not maximum_humidity_stpt_sch_id_b
+                    and maximum_humidity_stpt_sch_id_p
+                ):
+                    return FAIL_MSG_P
+                else:
+                    return ""
```

Comment:
There is no need for two ifs here. This function is called when RCT already determined it is failed.
So the message is only B or P.

Reply:
Good point!